src/usart/usart_frame.c

```diff
@@ -0,0 +1,347 @@
+     * byte of the header.
+     */
+    header_1 |= (data[0] << 8);
+    uint16_t cs = _ones_complement_sum(header_0, header_1);
+    cs = _ones_complement_sum(cs, _compute_checksum(data + 1, len - 1));
+    return cs;
+}
+
+static void _usart_send_frame_error(uint8_t error_code, uint32_t endpoint, struct queue* queue)
+{
+    uint8_t error_payload = endpoint;
+    usart_format_frame(error_code, &error_payload, 1, 0xFF, queue);
+}
+
+void usart_invalid_api_command(struct queue* queue, uint32_t src_endpoint)
+{
+    _usart_send_frame_error(USART_FRAME_ERROR_INVALID_CMD, src_endpoint, queue);
+}
+
+static void _usart_manage_full_rx_frame(void)
+{
+    // Check if this packet is correct
+    if (_usart_frame_parser_state.packet_size < 5) {
+        // Packet too short.
+        return;
+    }
+    // At the moment we only support version 1.
+    uint8_t version = _usart_frame_parser_state.buf[0];
+    if (version != 1) {
+        /*
+         * FUTURE: implement frame errors. For now
+         * we just drop bad frames.
+         */
+        return;
+    }
+    // Check the checksum, located in the last 2 bytes of the frame.
+    size_t payload_length = _usart_frame_parser_state.packet_size - 2;
+    uint16_t checksum = *((uint16_t*)(_usart_frame_parser_state.buf + payload_length));
+    uint16_t exp_checksum = _compute_checksum(_usart_frame_parser_state.buf, payload_length);
+    if (exp_checksum != checksum) {
+        /*
+         * FUTURE: implement frame errors. For now
+         * we just drop bad frames.
+         */
+        return;
+    }
+    uint8_t dst_endpoint = _usart_frame_parser_state.buf[1];
+    uint8_t u2f_command = _usart_frame_parser_state.buf[2];
+    if (!is_usart_endpoint_enabled(dst_endpoint)) {
+        _usart_send_frame_error(
+            USART_FRAME_ERROR_ENDPOINT_UNAVAILABLE, dst_endpoint, queue_hww_queue());
+        return;
+    }
+    bool can_process = usb_processing_enqueue(
+        usb_processing_hww(),
+        _usart_frame_parser_state.buf + 3,
+        payload_length - 3,
+        u2f_command,
+        dst_endpoint);
+    if (!can_process) {
+        _usart_send_frame_error(USART_FRAME_ERROR_ENDPOINT_BUSY, dst_endpoint, queue_hww_queue());
+    }
+}
+
+static void _usart_frame_packet_end(void)
+{
+    if (_usart_frame_parser_state.packet_size > 0) {
+        _usart_manage_full_rx_frame();
+    }
+    _usart_frame_reset_state();
+}
+
+static void _usart_frame_append_data_byte(uint8_t b)
+{
+    if (_usart_frame_parser_state.packet_size == USART_FRAME_MAX_DATA_LEN) {
+        // Error. Start looking for a new packet and discard the current one.
+        _usart_frame_reset_state();
+        return;
+    }
+    _usart_frame_parser_state.buf[_usart_frame_parser_state.packet_size] = b;
+    _usart_frame_parser_state.packet_size++;
+}
+
+static void _usart_frame_process_byte(uint8_t b)
+{
+    switch (_usart_frame_parser_state.state) {
+    case USART_FRAME_PARSE_IDLE:
```

Comment:
I don't think this will work if the packets come "back-to-back", i.e. with only 1 flag byte between the payloads. Is this by intention or? because I guess it would be quite simple to just start eating bytes if you are in IDLE and receive anything other than `FLAG_BYTE`.

Reply:
You could basically have a fall-thru to `USART_FRAME_PARSE_RX`. if you break in the conditional.

Reply:
That won't work for b2b packets. The correct way of moving forward here is to jump from end of packet straight into rx. `USART_FRAME_PARSE_IDLE` should only be entered when starting or after an error.

Reply:
Fixed.